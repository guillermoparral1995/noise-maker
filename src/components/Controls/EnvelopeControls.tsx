import React from 'react';
import Knob from '../shared/Knob';
import {
  updateAttack,
  updateDecay,
  updateRelease,
  updateSustain,
} from '../../store/actions';

const EnvelopeFilterControls = () => {
  return (
    <div>
      <Knob
        label="attack"
        defaultValue={0}
        action={updateAttack}
        min={0}
        max={5}
      ></Knob>
      <Knob
        label="decay"
        defaultValue={0}
        action={updateDecay}
        min={0}
        max={5}
      ></Knob>
      <Knob
        label="sustain"
        defaultValue={1}
        action={updateSustain}
        min={0}
        max={1}
      ></Knob>
      <Knob
        label="release"
        defaultValue={0}
        action={updateRelease}
        min={0}
        max={5}
      ></Knob>
    </div>
  );
};

export default EnvelopeFilterControls;
