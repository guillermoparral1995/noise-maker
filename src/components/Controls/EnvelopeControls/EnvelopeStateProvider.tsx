import React, { PropsWithChildren, useReducer } from 'react';
import initialState, { EnvelopeState } from './store/initialState';
import { ActionTypes } from './store/actions';
import reducer from './store/reducer';

interface EnvelopeStateProviderValue {
  state: EnvelopeState;
  dispatch: React.Dispatch<ActionTypes>;
}

export const envelopeStateContext =
  React.createContext<EnvelopeStateProviderValue>(undefined);

export const EnvelopeStateProvider = ({ children }: PropsWithChildren) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <envelopeStateContext.Provider value={{ state, dispatch }}>
      {children}
    </envelopeStateContext.Provider>
  );
};
