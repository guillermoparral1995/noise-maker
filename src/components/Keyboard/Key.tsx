import React, { useContext } from 'react';
import { audioContext } from '../../providers/AudioContextProvider';
import { stateContext } from '../../providers/StateProvider';
import { midiContext } from '../../providers/MIDIProvider';
import { NoteMessageEvent } from 'webmidi';

const Key = ({
  identifier,
  frequency,
}: {
  identifier: string;
  frequency: number;
}) => {
  const { context, output } = useContext(audioContext);
  const {
    state: { attack, decay, sustain, release, waveform },
  } = useContext(stateContext);
  const midiInput = useContext(midiContext);

  midiInput.addListener('noteon', (event: NoteMessageEvent) => {
    if (event.note.identifier === identifier) {
      play();
    }
  });

  midiInput.addListener('noteoff', (event: NoteMessageEvent) => {
    if (event.note.identifier === identifier) {
      stop();
    }
  });

  const oscillator = new OscillatorNode(context, { type: waveform, frequency });
  const envelope = new GainNode(context);
  envelope.connect(output);
  oscillator.start(context.currentTime);

  const play = () => {
    envelope.gain.setValueAtTime(0, context.currentTime);
    oscillator.connect(envelope);
    envelope.gain.linearRampToValueAtTime(
      1,
      context.currentTime + attack + 0.01,
    );
    envelope.gain.linearRampToValueAtTime(
      sustain + 0.01,
      context.currentTime + attack + decay,
    );
  };

  const stop = () => {
    envelope.gain.cancelAndHoldAtTime(context.currentTime);
    envelope.gain.exponentialRampToValueAtTime(
      0.01,
      context.currentTime + release,
    );
    setTimeout(
      () => {
        oscillator.disconnect();
      },
      (context.currentTime + release) * 1000 + 10,
    );
  };

  return (
    <>
      <button type="button" onMouseDown={play} onMouseUp={stop}>
        {identifier}
      </button>
    </>
  );
};

export default Key;
