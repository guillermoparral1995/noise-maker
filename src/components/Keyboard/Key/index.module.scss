.key {
  border: 1px solid black;
  border-bottom-left-radius: 2px;
  border-bottom-right-radius: 2px;
  width: 4em;

  &.white {
    &.pressed {
      background-color: gray;
    }
    background-color: white;
    box-shadow: 2px 2px 2px black;
    height: 13em;
  }

  &.black {
    &.pressed {
      background-color: gray;
    }
    background-color: black;
    border-color: white;
    border-top-color: black;

    box-shadow: 2px 2px 2px black;
    color: white;
    height: 7.6em;
    position: absolute;
    z-index: 1;

    &.Csharp3 {
      left: 2em;
    }

    &.Dsharp3 {
      left: 6em;
    }

    &.Fsharp3 {
      left: 14em;
    }

    &.Gsharp3 {
      left: 18em;
    }

    &.Asharp3 {
      left: 22em;
    }

    &.Csharp4 {
      left: 30em;
    }

    &.Dsharp4 {
      left: 34em;
    }

    &.Fsharp4 {
      left: 42em;
    }

    &.Gsharp4 {
      left: 46em;
    }

    &.Asharp4 {
      left: 50em;
    }
  }
}
