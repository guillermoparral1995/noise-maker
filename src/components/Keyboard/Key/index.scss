.key {
    border: 1px solid black;
    border-bottom-left-radius: 2px;
    border-bottom-right-radius: 2px;
    width: 25px;

    &.white {
        &.pressed {
            background-color: gray;
        }
        background-color: white;
        box-shadow: 2px 2px 2px black;
        height: 100px;
    }
    
    &.black {
        &.pressed {
            background-color: gray;
        }
        background-color: black;
        border-color: white;
        border-top-color: black;

        box-shadow: 2px 2px 2px black;
        color: white;
        height: 60px;
        position: absolute;
        z-index: 1;

        &.Csharp3 {
            left: 12px;
        }

        &.Dsharp3 {
            left: 37px;
        }

        &.Fsharp3 {
            left: 87px;
        }

        &.Gsharp3 {
            left: 112px;
        }

        &.Asharp3 {
            left: 137px;
        }

        &.Csharp4 {
            left: 187px;
        }

        &.Dsharp4 {
            left: 212px;
        }

        &.Fsharp4 {
            left: 262px;
        }

        &.Gsharp4 {
            left: 287px;
        }

        &.Asharp4 {
            left: 312px;
        }
    }
}
