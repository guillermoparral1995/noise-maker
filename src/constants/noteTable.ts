export interface NoteValue {
  frequency: number;
  mapping: string;
}

type NoteTable = Record<Notes, NoteValue>;

export enum Notes {
  C3 = 'C3',
  CSharp3 = 'C#3',
  D3 = 'D3',
  DSharp3 = 'D#3',
  E3 = 'E3',
  F3 = 'F3',
  FSharp3 = 'F#3',
  G3 = 'G3',
  GSharp3 = 'G#3',
  A3 = 'A3',
  ASharp3 = 'A#3',
  B3 = 'B3',
  C4 = 'C4',
  CSharp4 = 'C#4',
  D4 = 'D4',
  DSharp4 = 'D#4',
  E4 = 'E4',
  F4 = 'F4',
  FSharp4 = 'F#4',
  G4 = 'G4',
  GSharp4 = 'G#4',
  A4 = 'A4',
  ASharp4 = 'A#4',
  B4 = 'B4',
  C5 = 'C5',
}

const noteTable: NoteTable = {
  [Notes.C3]: {
    frequency: 130.81,
    mapping: 'q',
  },
  [Notes.CSharp3]: {
    frequency: 138.59,
    mapping: '2',
  },
  [Notes.D3]: {
    frequency: 146.83,
    mapping: 'w',
  },
  [Notes.DSharp3]: {
    frequency: 155.56,
    mapping: '3',
  },
  [Notes.E3]: {
    frequency: 164.81,
    mapping: 'e',
  },
  [Notes.F3]: {
    frequency: 174.61,
    mapping: 'r',
  },
  [Notes.FSharp3]: {
    frequency: 185,
    mapping: '5',
  },
  [Notes.G3]: {
    frequency: 196,
    mapping: 't',
  },
  [Notes.GSharp3]: {
    frequency: 207.65,
    mapping: '6',
  },
  [Notes.A3]: {
    frequency: 220,
    mapping: 'y',
  },
  [Notes.ASharp3]: {
    frequency: 233.08,
    mapping: '7',
  },
  [Notes.B3]: {
    frequency: 246.94,
    mapping: 'u',
  },
  [Notes.C4]: {
    frequency: 261.63,
    mapping: 'z',
  },
  [Notes.CSharp4]: {
    frequency: 277.18,
    mapping: 's',
  },
  [Notes.D4]: {
    frequency: 293.66,
    mapping: 'x',
  },
  [Notes.DSharp4]: {
    frequency: 311.13,
    mapping: 'd',
  },
  [Notes.E4]: {
    frequency: 329.63,
    mapping: 'c',
  },
  [Notes.F4]: {
    frequency: 349.23,
    mapping: 'v',
  },
  [Notes.FSharp4]: {
    frequency: 369.99,
    mapping: 'g',
  },
  [Notes.G4]: {
    frequency: 392,
    mapping: 'b',
  },
  [Notes.GSharp4]: {
    frequency: 415.3,
    mapping: 'h',
  },
  [Notes.A4]: {
    frequency: 440,
    mapping: 'n',
  },
  [Notes.ASharp4]: {
    frequency: 466.16,
    mapping: 'j',
  },
  [Notes.B4]: {
    frequency: 493.88,
    mapping: 'm',
  },
  [Notes.C5]: {
    frequency: 523.25,
    mapping: ',',
  },
};

export default noteTable;
