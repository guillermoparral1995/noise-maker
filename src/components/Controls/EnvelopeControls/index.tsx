import useAddMidiListeners from '@hooks/useAddMidiListeners';
import { Knobs } from '@types';
import React, { useContext } from 'react';
import ControlsRow from '../../shared/ControlsRow';
import Knob from '../../shared/Knob';
import { envelopeStateContext } from './EnvelopeStateProvider';

const EnvelopeControls = () => {
  const { state, dispatch } = useContext(envelopeStateContext);
  useAddMidiListeners(
    [Knobs.ATTACK, Knobs.DECAY, Knobs.SUSTAIN, Knobs.RELEASE],
    dispatch,
  );

  return (
    <>
      <h3>ADSR Envelope</h3>
      <ControlsRow>
        <Knob id={Knobs.ATTACK} value={state.attack} dispatch={dispatch}></Knob>
        <Knob id={Knobs.DECAY} value={state.decay} dispatch={dispatch}></Knob>
        <Knob
          id={Knobs.SUSTAIN}
          value={state.sustain}
          dispatch={dispatch}
        ></Knob>
        <Knob
          id={Knobs.RELEASE}
          value={state.release}
          dispatch={dispatch}
        ></Knob>
      </ControlsRow>
    </>
  );
};

export default EnvelopeControls;
