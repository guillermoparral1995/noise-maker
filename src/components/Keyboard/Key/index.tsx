import React, { useContext, useEffect, useMemo, useRef } from 'react';
import { NoteMessageEvent } from 'webmidi';
import { audioContext } from '../../../providers/AudioContextProvider';
import { midiContext } from '../../../providers/MIDIProvider';
import { envelopeStateContext } from '../../Controls/EnvelopeControls/EnvelopeStateProvider';
import './index.scss';
import useConnectLFOTargets from '../../../hooks/useConnectLFOTargets';
import { Knobs } from '../../../types';

const Key = ({
  identifier,
  frequency,
}: {
  identifier: string;
  frequency: number;
}) => {
  const keyRef = useRef<HTMLButtonElement>(null);
  const { context, output, lfo1, lfo2 } = useContext(audioContext);
  const {
    state: { attack, decay, sustain, release, detune, waveform },
  } = useContext(envelopeStateContext);
  const midiInput = useContext(midiContext);
  let releaseTimeout: NodeJS.Timeout;

  const oscillator = useMemo(
    () => new OscillatorNode(context, { type: waveform, frequency }),
    [],
  );
  const envelope = useMemo(() => new GainNode(context), []);
  oscillator.type = waveform;
  oscillator.detune.value = detune;

  useEffect(() => {
    oscillator.start(context.currentTime);
    envelope.connect(output);
    return () => oscillator.disconnect();
  }, []);

  useConnectLFOTargets(lfo1, [
    { knob: Knobs.DETUNE, param: oscillator.detune },
  ]);
  useConnectLFOTargets(lfo2, [
    { knob: Knobs.DETUNE, param: oscillator.detune },
  ]);

  useEffect(() => {
    if (midiInput) {
      midiInput.addListener('noteon', (event: NoteMessageEvent) => {
        if (event.note.identifier === identifier) {
          play();
        }
      });

      midiInput.addListener('noteoff', (event: NoteMessageEvent) => {
        if (event.note.identifier === identifier) {
          stop();
        }
      });
    }
    return () => {
      if (midiInput) {
        midiInput.removeListener('noteon');
        midiInput.removeListener('noteoff');
      }
    };
  });

  const play = () => {
    if (keyRef.current) {
      keyRef.current.classList.add('pressed');
    }
    // clear all pending scheduled events for envelope
    envelope.gain.cancelAndHoldAtTime(context.currentTime);
    clearTimeout(releaseTimeout);

    // reset envelope and connect oscillator
    envelope.gain.setValueAtTime(0, context.currentTime);
    oscillator.connect(envelope);

    // ramp up to attack value
    envelope.gain.linearRampToValueAtTime(1, context.currentTime + attack);

    // ramp down to sustain value in decay time
    envelope.gain.linearRampToValueAtTime(
      sustain,
      context.currentTime + attack + decay,
    );
  };

  const stop = () => {
    if (keyRef.current) {
      keyRef.current.classList.remove('pressed');
    }
    // clear pending scheduled evenets for envelope
    envelope.gain.cancelAndHoldAtTime(context.currentTime);

    // ramp down to 0 in release time
    envelope.gain.exponentialRampToValueAtTime(
      0.001,
      context.currentTime + release,
    );

    // set timeout for disconnecting oscillator after release
    releaseTimeout = setTimeout(
      () => {
        oscillator.disconnect();
      },
      (context.currentTime + release) * 1000 + 10,
    );
  };

  return (
    <button
      ref={keyRef}
      className={`key ${
        identifier.includes('#') ? 'black' : 'white'
      } ${identifier.replace('#', 'sharp')}`}
      type="button"
      onMouseDown={play}
      onMouseUp={stop}
    ></button>
  );
};

export default Key;
