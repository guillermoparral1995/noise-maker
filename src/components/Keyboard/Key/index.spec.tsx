import { cleanup, fireEvent, render } from '@testing-library/react';
import React from 'react';
import Key from '.';
import {
  AudioContextMock,
  AudioNodeMock,
  DelayNodeMock,
  GainNodeMock,
  OscillatorNodeMock,
  withMockedMIDIInput,
  withMockedMIDINoInput,
} from '../../../../__mocks__';
import noteTable, { Notes } from '../../../constants/noteTable';
import {
  AudioContextProvider,
  LFOMock,
} from '../../../providers/AudioContextProvider';
import { Knobs, LFO1Target, LFO2Target } from '../../../types';
import { EnvelopeStateProvider } from '../../Controls/EnvelopeControls/EnvelopeStateProvider';

const __mockGetInputByName = jest.fn();
jest.mock('webmidi', () => {
  return {
    ...jest.requireActual('webmidi'),
    WebMidi: {
      inputs: [],
      enable: jest.fn(),
      getInputByName: () => __mockGetInputByName(),
    },
  };
});

const checkSoundIsPlayed = (
  mockOscillator: OscillatorNodeMock,
  mockEnvelope: GainNodeMock,
) => {
  expect(mockEnvelope.gain.__mockCancelAndHoldAtTime).toHaveBeenCalledWith(
    1000,
  );
  expect(mockEnvelope.gain.__mockSetValueAtTime).toHaveBeenCalledWith(0, 1000);
  expect(clearTimeout).toHaveBeenCalled();
  expect(mockOscillator.__mockConnect).toHaveBeenCalledWith(mockEnvelope);
  expect(mockEnvelope.gain.__mockLinearRampToValueAtTime).toHaveBeenCalledWith(
    1,
    1000,
  );
};

const checkSoundIsStopped = (
  mockOscillator: OscillatorNodeMock,
  mockEnvelope: GainNodeMock,
) => {
  expect(mockEnvelope.gain.__mockCancelAndHoldAtTime).toHaveBeenCalledWith(
    1000,
  );
  expect(
    mockEnvelope.gain.__mockExponentialRampToValueAtTime,
  ).toHaveBeenCalledWith(0.001, 1000);
  expect(setTimeout).toHaveBeenCalledWith(
    expect.any(Function),
    1000 * 1000 + 10,
  );
  jest.runAllTimers();

  expect(mockOscillator.__mockDisconnect).toHaveBeenCalled();
};

describe('Key', () => {
  beforeAll(() => {
    jest.useFakeTimers();
    jest.spyOn(global, 'setTimeout');
    jest.spyOn(global, 'clearTimeout');
  });
  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
    jest.clearAllTimers();
  });

  const mockContext = new AudioContextMock();
  const mockLfo1: LFOMock<LFO1Target> = {
    target: 'off',
    output: new GainNodeMock(),
  };
  const mockLfo2: LFOMock<LFO2Target> = {
    target: 'off',
    output: new GainNodeMock(),
  };
  const mockOutput = new AudioNodeMock();
  const mockOscillator = new OscillatorNodeMock();
  const mockEnvelope = new GainNodeMock();
  const mockDelay = {
    node: new DelayNodeMock(),
  };

  it('should render correctly', async () => {
    const { findByTestId } = render(
      withMockedMIDINoInput(
        <AudioContextProvider
          __mocks={{
            context: mockContext,
            lfo1: mockLfo1,
            lfo2: mockLfo2,
            delay: mockDelay,
            output: mockOutput,
          }}
        >
          <EnvelopeStateProvider>
            <Key
              identifier={Notes.C3}
              frequency={noteTable[Notes.C3]}
              __mockOscillator={mockOscillator}
              __mockEnvelope={mockEnvelope}
            ></Key>
            ,
          </EnvelopeStateProvider>
        </AudioContextProvider>,
      ),
    );

    const key = await findByTestId(Notes.C3);
    expect(mockOscillator.__mockStart).toHaveBeenCalled();
    expect(mockEnvelope.__mockConnect).toHaveBeenCalledWith(mockOutput);
    expect(key).toMatchSnapshot();
  });

  it('should play sound on mouse down', async () => {
    const { findByTestId } = render(
      withMockedMIDINoInput(
        <AudioContextProvider
          __mocks={{
            context: mockContext,
            lfo1: mockLfo1,
            lfo2: mockLfo2,
            delay: mockDelay,
          }}
        >
          <EnvelopeStateProvider>
            <Key
              identifier={Notes.C3}
              frequency={noteTable[Notes.C3]}
              __mockOscillator={mockOscillator}
              __mockEnvelope={mockEnvelope}
            ></Key>
            ,
          </EnvelopeStateProvider>
        </AudioContextProvider>,
      ),
    );

    const key = await findByTestId(Notes.C3);
    fireEvent.mouseDown(key);
    checkSoundIsPlayed(mockOscillator, mockEnvelope);
    expect(key).toMatchSnapshot();
  });

  it('should play sound on MIDI press', async () => {
    const mockAddListener = jest.fn(
      (eventName: string, listener: (e: object) => void) => {
        if (eventName === 'noteon') {
          listener({ note: { identifier: Notes.C3 } });
        }
      },
    );

    __mockGetInputByName.mockReturnValue({
      addListener: mockAddListener,
      removeListener: jest.fn(),
    });

    const { findByTestId } = render(
      withMockedMIDIInput(
        <AudioContextProvider
          __mocks={{
            context: mockContext,
            lfo1: mockLfo1,
            lfo2: mockLfo2,
            delay: mockDelay,
          }}
        >
          <EnvelopeStateProvider>
            <Key
              identifier={Notes.C3}
              frequency={noteTable[Notes.C3]}
              __mockOscillator={mockOscillator}
              __mockEnvelope={mockEnvelope}
            ></Key>
            ,
          </EnvelopeStateProvider>
        </AudioContextProvider>,
      ),
    );

    const key = await findByTestId(Notes.C3);

    expect(mockAddListener).toHaveBeenCalledTimes(2);
    expect(mockAddListener).toHaveBeenLastCalledWith(
      'noteoff',
      expect.any(Function),
    );
    expect(mockAddListener).toHaveBeenNthCalledWith(
      1,
      'noteon',
      expect.any(Function),
    );

    checkSoundIsPlayed(mockOscillator, mockEnvelope);
    expect(key).toMatchSnapshot();
  });

  it('should stop sound on mouse up', async () => {
    const { findByTestId } = render(
      withMockedMIDINoInput(
        <AudioContextProvider
          __mocks={{
            context: mockContext,
            lfo1: mockLfo1,
            lfo2: mockLfo2,
            delay: mockDelay,
          }}
        >
          <EnvelopeStateProvider>
            <Key
              identifier={Notes.C3}
              frequency={noteTable[Notes.C3]}
              __mockOscillator={mockOscillator}
              __mockEnvelope={mockEnvelope}
            ></Key>
            ,
          </EnvelopeStateProvider>
        </AudioContextProvider>,
      ),
    );

    const key = await findByTestId(Notes.C3);
    fireEvent.mouseDown(key);
    fireEvent.mouseUp(key);

    checkSoundIsStopped(mockOscillator, mockEnvelope);
    expect(key).toMatchSnapshot();
  });

  it('should stop sound on MIDI unpress', async () => {
    const mockAddListener = jest.fn(
      (eventName: string, listener: (e: object) => void) => {
        if (eventName === 'noteoff') {
          listener({ note: { identifier: Notes.C3 } });
        }
      },
    );

    __mockGetInputByName.mockReturnValue({
      addListener: mockAddListener,
      removeListener: jest.fn(),
    });

    const { findByTestId } = render(
      withMockedMIDIInput(
        <AudioContextProvider
          __mocks={{
            context: mockContext,
            lfo1: mockLfo1,
            lfo2: mockLfo2,
            delay: mockDelay,
          }}
        >
          <EnvelopeStateProvider>
            <Key
              identifier={Notes.C3}
              frequency={noteTable[Notes.C3]}
              __mockOscillator={mockOscillator}
              __mockEnvelope={mockEnvelope}
            ></Key>
            ,
          </EnvelopeStateProvider>
        </AudioContextProvider>,
      ),
    );

    const key = await findByTestId(Notes.C3);
    checkSoundIsStopped(mockOscillator, mockEnvelope);
    expect(key).toMatchSnapshot();
  });

  it('should connect lfo1 with oscillator detune', async () => {
    const mockLfo1: LFOMock<LFO1Target> = {
      target: Knobs.DETUNE,
      output: new GainNodeMock(),
    };
    const { findByTestId } = render(
      withMockedMIDIInput(
        <AudioContextProvider
          __mocks={{
            context: mockContext,
            lfo1: mockLfo1,
            lfo2: mockLfo2,
            delay: mockDelay,
          }}
        >
          <EnvelopeStateProvider>
            <Key
              identifier={Notes.C3}
              frequency={noteTable[Notes.C3]}
              __mockOscillator={mockOscillator}
            ></Key>
            ,
          </EnvelopeStateProvider>
        </AudioContextProvider>,
      ),
    );
    await findByTestId(Notes.C3);

    expect(mockLfo1.output.__mockConnect).toHaveBeenCalledWith(
      mockOscillator.detune,
    );
    expect(mockLfo2.output.__mockConnect).not.toHaveBeenCalled();
  });

  it('should connect lfo2 with oscillator detune', async () => {
    const mockLfo2: LFOMock<LFO2Target> = {
      target: Knobs.DETUNE,
      output: new GainNodeMock(),
    };
    const { findByTestId } = render(
      withMockedMIDIInput(
        <AudioContextProvider
          __mocks={{
            context: mockContext,
            lfo1: mockLfo1,
            lfo2: mockLfo2,
            delay: mockDelay,
          }}
        >
          <EnvelopeStateProvider>
            <Key
              identifier={Notes.C3}
              frequency={noteTable[Notes.C3]}
              __mockOscillator={mockOscillator}
            ></Key>
            ,
          </EnvelopeStateProvider>
        </AudioContextProvider>,
      ),
    );
    await findByTestId(Notes.C3);

    expect(mockLfo2.output.__mockConnect).toHaveBeenCalledWith(
      mockOscillator.detune,
    );
    expect(mockLfo1.output.__mockConnect).not.toHaveBeenCalledWith();
  });
});
