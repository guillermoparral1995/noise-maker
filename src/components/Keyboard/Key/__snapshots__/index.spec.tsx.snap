// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Key should play sound on MIDI press 1`] = `
<button
  class="key white pressed"
  data-testid="C3"
  tabindex="-1"
  type="button"
/>
`;

exports[`Key should play sound on mouse down 1`] = `
<button
  class="key white pressed"
  data-testid="C3"
  tabindex="-1"
  type="button"
/>
`;

exports[`Key should render correctly 1`] = `
<button
  class="key white"
  data-testid="C3"
  tabindex="-1"
  type="button"
/>
`;

exports[`Key should stop sound on MIDI unpress 1`] = `
<button
  class="key white"
  data-testid="C3"
  tabindex="-1"
  type="button"
/>
`;

exports[`Key should stop sound on mouse up 1`] = `
<button
  class="key white"
  data-testid="C3"
  tabindex="-1"
  type="button"
/>
`;
